File: be1-go/message/messagedata/federation_request_challenge.go
package messagedata

// FederationRequestChallenge defines a message data
type FederationRequestChallenge struct {
	Object string `json:"object"`
	Action string `json:"action"`

	// Timestamp is a Unix timestamp
	Timestamp int64 `json:"timestamp"`
}

// GetObject implements MessageData
func (FederationRequestChallenge) GetObject() string {
	return FederationObject
}

// GetAction implements MessageData
func (FederationRequestChallenge) GetAction() string {
	return FederationActionRequestChallenge
}

// NewEmpty implements MessageData
func (FederationRequestChallenge) NewEmpty() MessageData {
	return &FederationRequestChallenge{}
}

File: be2-scala/src/main/scala/ch/epfl/pop/model/network/method/message/data/federation/FederationRequestChallenge.scala
package ch.epfl.pop.model.network.method.message.data.federation

import ch.epfl.pop.json.MessageDataProtocol.FederationRequestChallengeFormat
import ch.epfl.pop.model.network.Parsable
import ch.epfl.pop.model.network.method.message.data.{ActionType, MessageData, ObjectType}
import ch.epfl.pop.model.objects.Timestamp
import spray.json.*

final case class FederationRequestChallenge(
    timestamp: Timestamp
) extends MessageData {
  override val _object: ObjectType = ObjectType.federation
  override val action: ActionType = ActionType.challenge_request
}

object FederationRequestChallenge extends Parsable {
  def apply(timestamp: Timestamp): FederationRequestChallenge = {
    new FederationRequestChallenge(timestamp)
  }

  override def buildFromJson(payload: String): FederationRequestChallenge = payload.parseJson.asJsObject.convertTo[FederationRequestChallenge]
}
